Publishing Wizard improvements #5497

edloidas opened the pull request:
Replaced a single warning line with more complex issues state bar, that will be universal among multiple dialogs.
Replaced issues state bar in publish dialog.
Made Content resource return number of not publishable content.
Made Content resource return number of not publishable content.
Most of the markChecking() calls are made for smoother UI transitions.

edloidas committed:
Open Publishing wizard after content is marked as ready #5413

edloidas committed:
Archive dialog improvements #5432

alansemenov committed:
Enable apps on projects

anatol-sialitski committed:
Content Variant A/B testing

pmi committed:
Simplify version comparison for a published content #5412

ashklianko committed:
Enable adding a new content directly from Content Selector #5436

edloidas committed:
Multi-inheritance in layers #5414

alansemenov commented:
@vbradnitski please review the Java part

edloidas committed:
Publishing Wizard improvements #5497

Replaced a single warning line with more complex issues state bar, that will be universal among multiple dialogs.
Replaced issues state bar in publish dialog.
Made Content resource return number of not publishable content.
Made Content resource return number of not publishable content.
Most of the markChecking() calls are made for smoother UI transitions.
Made it possible to make entry non-interactive by hiding the button in case it is no longer possible to interact with it, like in cases of non-excludable items from publish.

ashklianko commented:
When opening a publishing dialog with many items multiple items it may show for a short moment that all checks are passed and green:

https://user-images.githubusercontent.com/4255706/206406227-c5369af9-947a-48c5-97b0-9e5b34550bf9.mp4

ashklianko commented:
Don't we need the same status bar in ContentUnpublishDIalog?

ashklianko commented:
Looks good @edloidas !

ashklianko commented on modules/lib/src/main/resources/assets/js/app/dialog/DialogErrorsStateBar.ts:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    protected updateResolvedState(): void {
+        const isResolved = !this.errorsEntries.some(entry => entry.isActive());
```
`const isResolved: boolean = !this.errorsEntries.some((entry: DialogErrorStateEntry) => entry.isActive());`
I believe we have agreed long ago to always declare types on variables to make it easier to read, even though some types can be inferred, however in this case it requires to go and check what is entry's type

edloidas replied:
Oh boy, we haven't even got started. Guess we'll have to solve this with a one-on-one fight with knifes.

ashklianko commented on modules/lib/src/main/resources/assets/js/app/dialog/DialogErrorsStateBar.ts:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    setEnabled(enabled: boolean): void {
+        this.errorsEntries.forEach(entry => entry.setActionEnabled(enabled));
```
Same here: `entry` is missing type

ashklianko commented on modules/lib/src/main/resources/assets/js/app/dialog/DialogErrorsStateBar.ts:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    markChecking(checking: boolean): void {
+        this.errorsEntries.forEach(entry => entry.markChecking(checking));
```
Same here: `entry` is missing type

ashklianko commented on modules/lib/src/main/resources/assets/js/app/dialog/DialogErrorsStateBar.ts:
```diff
@@ -0,0 +1,152 @@
+    toggleHideIfResolved(hideIfResolved: boolean): void {
+        if (this.hideIfResolved !== hideIfResolved) {
+            this.hideIfResolved = hideIfResolved;
+            const isResolved = !this.errorsEntries.some(entry => entry.isActive());
```
Types are missing

alansemenov commented:
> Don't we need the same status bar in ContentUnpublishDIalog?

Correct, but this will be done in a separate task.

edloidas committed:
Publishing Wizard improvements #5497

Added more types.
Prevented events from triggering handlers after reset, or when containers are not visible.

edloidas commented:
Fixed the errors bar state problem.
It appears, that the initial problem behind the invalid errors bar was there for quite a long time and by using the debounced onLoadingFinished handlers just made it worse.
Because of the unholy mix of multiple debounced handlers, events and promises, that are used simultaneously, the handlers were triggered when they must not (visually this problem was affecting loading mask and controls lock state earlier).

Closing the dialog was triggering unnecessary delayed notifications, that had effect if you were quick enough to open it before the close callback were actually done running. Or you can just open the dialog, close it, and open again (in case the background checks will take long enough) to get double the notifications that will break the UI.

If we are to use it somewhere else in the future or add more functionality, I would recommend to write the publish processor from scratch if there'll be time.

alansemenov commented:
Moved to https://github.com/enonic/app-contentstudio/pull/5562
